nucypher/blockchain/eth/interfaces.py

```diff
@@ -370,11 +368,11 @@ def sync(self, emitter=None) -> None:
         return
 
     @property
-    def provider(self) -> Union[IPCProvider, WebsocketProvider, HTTPProvider]:
+    def provider(self) -> BaseProvider:
         return self._provider
 
     def _attach_provider(self,
-                         provider: Web3Providers = None,
+                         provider: BaseProvider = None,
```

Comment:
`Optional`? Also, how come the default `provider` value in `__init__()` is `NO_BLOCKCHAIN_CONNECTION`, but here it's just `None`?

Reply:
I think the parameters in the function were really added to facilitate for testing in test_web3_clients.py (previous code, see Mock Clients). In the default case i.e. internal call, `self.provider` and `self.provider_uri` are passed. However, in the testing case, only a provider is passed.

My goal was to get rid of `Web3Providers` type because in our code it never mattered what the underlying Union types (Websocket, HTTP, IPC) was, we just pass the object around, and web3 eventually uses it.

Reply:
Makes sense - but yeah make it `Optional[BaseProvider]` if it can be `None`

Reply:
✔️